### PR TITLE
Add testing cluster

### PR DESCRIPTION
**Pull request by n8maninger**
Daemon for spinning up a cluster of nodes and connecting them to a local
test network. Useful for local e2e testing.

## Updating Daemons

The cluster daemon uses go.mod to manage the daemon dependency. To test a specific
version of a daemon against a cluster, update the go.mod file to point to the
desired commit.

```
go get go.sia.tech/hostd@asdasd
```

## Usage

```bash
$ go run ./cmd/clusterd -renterd=1 --hostd=3
```

Once the cluster is up, the cluster API will be available to interact with the
testnet.

Each node has a unique api address to interact with its respective daemon.

CTRL-C or SIGINT will shut down the cluster and cleanup the resources

### `[GET] /nodes` 
Lists all running nodes in the cluster

**Response Body**
```json
[
        {
                "id": "facaba60da2e71bb",
                "type": "renterd",
                "apiAddress": "http://[::]:53873/api",
                "password": "sia is cool",
                "walletAddress": "addr:1ca3bfe60a50fe3e700a67fae7c9670446849a0177b0da0378399c6d6ca9cb13dfcb3d084c2d"
        },
        {
                "id": "eabf85c452cf4bca",
                "type": "walletd",
                "apiAddress": "http://[::]:53874/api",
                "password": "sia is cool",
                "walletAddress": "addr:000000000000000000000000000000000000000000000000000000000000000089eb0d6a8a69"
        },
        {
                "id": "1c01bc97b794c8d7",
                "type": "hostd",
                "apiAddress": "http://[::]:53864/api",
                "password": "sia is cool",
                "walletAddress": "addr:862b5ca325b4c77ffe861a892e65f48bbcb43f2287b2dfe8d7c70459659f7367a2ffbcf62de9"
        },
        {
                "id": "31b3a65f8a1842d2",
                "type": "hostd",
                "apiAddress": "http://[::]:53867/api",
                "password": "sia is cool",
                "walletAddress": "addr:b906d603430c7e814a107910d8029f931da96a6f2afef6e6dfa93b488e9b337ae29d9f195d23"
        },
        {
                "id": "712e5c22050c0700",
                "type": "hostd",
                "apiAddress": "http://[::]:53870/api",
                "password": "sia is cool",
                "walletAddress": "addr:42fd84ddb08a96105c41b31414913a1a3e09a94046d8444f65398e9b96a814a7708027a323f7"
        }
]
```

### `POST /mine` 

Mines a preset number of blocks on the testnet

**Request Body**
```json
{
	"blocks": 10
}
```

### CODE REVIEW

File: internal/cluster/cmd/clusterd/main.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		network  string
+
+		siafundAddr string
+
+		renterdCount int
+		hostdCount   int
+		walletdCount int
+	)
+
+	flag.StringVar(&dir, "dir", "", "directory to store renter data")
```
Comment (alexfreska) [inline internal/cluster/cmd/clusterd/main.go:41]: This `dir` stores all hosts and wallets data too right?

**Review by alexfreska — COMMENTED** [review]
Started testing this out, its great. The only thing I hit is CORS issues from the UI instance, could we add some middleware to drop CORS restrictions from all the nodes?

**Review by alexfreska — COMMENTED** [review]
Screenshot 2024-08-19 at 1 54 17 PM
I am also having trouble gettting the hosts usable from renterd perspective, I see the hosts are set to AcceptingContracts true by default, and they show as not gouging, is there anything I likely need to set?

**Comment by n8maninger** [comment]
> Screenshot 2024-08-19 at 1 54 17 PM I am also having trouble gettting the hosts usable from renterd perspective, I see the hosts are set to AcceptingContracts true by default, and they show as not gouging, is there anything I likely need to set?

Is there a log on why?

**Comment by n8maninger** [comment]
@alexfreska the renter node should be automatically forming contracts with any host nodes and I've added the CORS headers to all nodes in the cluster.

**Comment by alexfreska** [comment]
> @alexfreska the renter node should be automatically forming contracts with any host nodes and I've added the CORS headers to all nodes in the cluster. I also added an endpoint to proxy requests to multiple nodes in the cluster, can be useful when debugging or modifying hosts.

@n8maninger I've pulled and built the latest, I am seeing some CORS errors:
Screenshot 2024-08-20 at 11 01 56 AM

**Comment by n8maninger** [comment]
@alexfreska sorry, should be fixed
Screenshot 2024-08-20 at 8 12 31 AM

**Comment by alexfreska** [comment]
@n8maninger this is working great. As I migrate the tests I may ask for help adding more parameters, but lets get this merged. Looks like 2 issues, a commit lint, but also the go upgrades? may have made the sdk build unhappy:
Screenshot 2024-08-22 at 3 14 19 PM

**Review by alexfreska — APPROVED** [review]
@n8maninger this is working great. As I migrate the tests I may ask for help adding more parameters, but lets get this merged!

Looks like 2 issues, a commit lint, but also the go upgrades? may have made the sdk build unhappy:
Screenshot 2024-08-22 at 3 14 19 PM

**Comment by n8maninger** [comment]
@alexfreska the original error was because I inadvertently forced a `core` dependency update due to the existince of a `go.work` file. I refactored this PR to be as minimal as possible by removing the go.work file and updating some paths. Web is using an ancient version of core and should eventually update its dependency.

**Review by alexfreska — APPROVED** [review]

**Pull request merged**